symfit/core/fit.py

```diff
@@ -133,23 +170,98 @@ def _init_from_dict(self, model_dict):
+        sort_func = lambda symbol: symbol.name
+        self.model_dict = OrderedDict(sorted(model_dict.items(),
+                                             key=lambda i: sort_func(i[0])))
+        # Everything at the bottom of the toposort is independent, at the top
+        # dependent, and the rest interdependent.
+        ordered = list(toposort(self.connectivity_mapping))
+        independent = ordered.pop(0)
+        self.dependent_vars = sorted(ordered.pop(-1), key=sort_func)
+        self.interdependent_vars = sorted(
+            [item for items in ordered for item in items],
+            key=sort_func
+        )
+        # `independent` contains both params and vars, needs to be separated
+        self.params = sorted(
+            [s for s in independent if isinstance(s, Parameter)],
+            key=sort_func
+        )
+        self.independent_vars = sorted(
+            [s for s in independent
+             if not isinstance(s, Parameter) and not s in self],
+            key=sort_func
+        )
 
-        # Extract all the params and vars as a sorted, unique list.
-        expressions = model_dict.values()
-        _params, self.independent_vars = set([]), set([])
-        for expression in expressions:
-            vars, params = seperate_symbols(expression)
-            _params.update(params)
-            self.independent_vars.update(vars)
-        # Although unique now, params and vars should be sorted alphabetically to prevent ambiguity
-        self.params = sorted(_params, key=sort_func)
-        self.independent_vars = sorted(self.independent_vars, key=sort_func)
+        # Make Variable object corresponding to each depedent var.
+        self.sigmas = {var: Variable(name='sigma_{}'.format(var.name))
+                       for var in self.dependent_vars}
 
-        # Make Variable object corresponding to each var.
-        self.sigmas = {var: Variable(name='sigma_{}'.format(var.name)) for var in self.dependent_vars}
+    @cached_property
+    def vars_as_functions(self):
+        """
+        :return: Turn the keys of this model into :mod:`~sympy.Function`
+            objects. This is done recursively so the chain rule can be applied
+            correctly. This is done on the basis of `connectivity_mapping`.
+
+            Example: for ``{y: a * x}`` this returns ``{y: y(x, a)}``.
+        """
+        functions = {}
+        # vars first, then params, and alphabetically within each group
+        key = lambda arg: [isinstance(arg, Parameter), str(arg)]
+        for symbol in self.ordered_symbols:
+            if symbol in self.connectivity_mapping:
+                connections = self.connectivity_mapping[symbol]
+                # Replace the connection by it's function if possible
+                connections = [functions.get(connection, connection)
+                               for connection in connections]
+                connections = sorted(connections, key=key)
```

Comment:
This is melting my brain. `connections` is a list containing either Functions, Variables, or Parameters, right? How do Functions get sorted compared to Variables in this story?

Reply:
I'll add some comments to the code ;). (And do some renaming of variables) Basically, `functions` will map vars to functions, right? And since we walk the tolopological tree in order, I substitute every variable in connections by its corresponding function if it is already present in `function`. Due to the order, this is guaranteed to work. So e.g. if I have
```python
model_dict = {z: y, y: x}
```
then the order will be `[x, y, z]`. Then, since `x` is not a key in connectivity mapping, it gets ignored. For `y`, we add `{y: Function('y')(x)}` to `functions`. For `z`, we add `{z: Function('z')(Function('y')(x))}`, which upon printing looks a lot more readable:
```python
{y: y(x), z: z(y(x))}
```

Reply:
Ah ok. I think it finally did "click". Still requires more comments/variable names though.